### PR TITLE
[Kernel] Support beforeOrAt and atOrAfter semantics for getting a commit version from a timestamp

### PR DESCRIPTION
**Pull request by allisonport-db**
#### Which Delta project/connector is this regarding?

- [ ] Spark
- [ ] Standalone
- [ ] Flink
- [X] Kernel
- [ ] Other (fill in here)

## Description

Adds the functions `getVersionBeforeOrAtTimestamp` and `getVersionAtOrAfterTimestamp` to `TableImpl` for use with the `getChangesByVersion` API to enable querying changes between two timestamps. 

## How was this patch tested?

Adds unit tests.

### CODE REVIEW

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java [hunk 1]
```diff
@@ -120,6 +120,78 @@ protected Path getLogPath() {
...
+   * </ul>.
+   *
+   * @param millisSinceEpochUTC the number of milliseconds since midnight, January 1, 1970 UTC
+   * @return latest commit that happened at or before {@code timestamp}.
+   * @throws KernelException if the timestamp is more than the timestamp of any committed
+   *                         version
+   * @throws TableNotFoundException if no delta table is found
+   */
+  public long getVersionAtOrAfterTimestamp(Engine engine, long millisSinceEpochUTC) {
+      DeltaHistoryManager.Commit commit = DeltaHistoryManager.getActiveCommitAtTimestamp(engine, getLogPath(), millisSinceEpochUTC, false, /* mustBeRecreatable */
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java]: formatting, I'm gonna guess the javastyle complains about this

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/DeltaHistoryManager.java [hunk 2]
```diff
@@ -40,44 +41,66 @@ private DeltaHistoryManager() {}
...
+   *                         canReturnLastCommit is false
    * @throws TableNotFoundException when there is no Delta table at the given path
    */
-  public static long getActiveCommitAtTimestamp(Engine engine, Path logPath, long timestamp)
-      throws TableNotFoundException {
+  public static Commit getActiveCommitAtTimestamp(
+      Engine engine,
+      Path logPath,
+      long timestamp,
+      boolean mustBeRecreatable,
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/DeltaHistoryManager.java:68]: this is only ever invoked with `mustBeRecreatable =false`. do we ever expect it to be invoked with `mustBeRecreatable=true`? if not, can we remove this param and then cut down our testing space in half?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java [hunk 3]
```diff
@@ -120,6 +120,78 @@ protected Path getLogPath() {
...
+          false, /* canReturnLastCommit */
+          // e.g. we give time T-1 and first commit has time T, then we DO want that earliest
+          // commit
+          true /* canReturnEarliestCommit */);
+
+      if (commit.timestamp >= millisSinceEpochUTC) {
+          return commit.version;
+      } else {
+          // this commit.timestamp is before the input timestamp. if this is the last commit, then the
+          // input timestamp is after the last commit and `getActiveCommitAtTime` would have thrown
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java]: `getActiveCommitAtTimestamp`

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/DeltaHistoryManager.java [hunk 4]
```diff
@@ -242,10 +289,10 @@ private static Optional<Commit> lastCommitBeforeOrAtTimestamp(
     return Optional.ofNullable((i < 0) ? null : commits.get(i));
   }
 
-  private static class Commit {
+  public static class Commit {
 
-    final long version;
-    final long timestamp;
+    public final long version;
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/DeltaHistoryManager.java:294]: can we use getters instead?

File: kernel/kernel-api/src/test/scala/io/delta/kernel/internal/DeltaHistoryManagerSuite.scala [hunk 5]
```diff
@@ -29,25 +29,52 @@ class DeltaHistoryManagerSuite extends AnyFunSuite with MockFileSystemClientUtil
...
+      timestamp,
+      mustBeRecreatable,
+      canReturnLastCommit,
+      canReturnEarliestCommit
     )
-    assert(activeCommit == expectedVersion,
+    assert(activeCommit.version == expectedVersion,
       s"Expected version $expectedVersion but got $activeCommit for timestamp=$timestamp")
+
+    if (mustBeRecreatable) {
```
Comment (scottsand-db) [inline kernel/kernel-api/src/test/scala/io/delta/kernel/internal/DeltaHistoryManagerSuite.scala:47]: As per my previous comment, I'm just curious why we need to test this extra case? What's a future use case where we want `mustBeRecreatable` to be true?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java (hunk 1 above)
Comment (allisonport-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java]: yeah for some reason merging master messed with the formatting... I've fixed it

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/DeltaHistoryManager.java (hunk 2 above)
Comment (allisonport-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/DeltaHistoryManager.java:68]: We use it with `mustBeRecreatable=true` for snapshot time travel queries

File: kernel/kernel-api/src/test/scala/io/delta/kernel/internal/DeltaHistoryManagerSuite.scala (hunk 5 above)
Comment (allisonport-db) [reply kernel/kernel-api/src/test/scala/io/delta/kernel/internal/DeltaHistoryManagerSuite.scala:47]: answered above

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/DeltaHistoryManager.java (hunk 2 above)
Comment (scottsand-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/DeltaHistoryManager.java:68]: Ah, missed that! Thanks!

**Review by scottsand-db — APPROVED** [review]

**Pull request merged**